controller/lib/core/blower_fsm.cpp

```diff
@@ -15,182 +15,12 @@ limitations under the License.
-  static Duration inspire_duration(const VentParams &params) {
-    float t =
-        60.f / static_cast<float>(params.breaths_per_min); // secs per breath
-    float r = params.inspiratory_expiratory_ratio; // I:E
-    return seconds(t * r / (1 + r));
-  }
-  static Duration expire_duration(const VentParams &params) {
-    float t =
-        60.f / static_cast<float>(params.breaths_per_min); // secs per breath
-    float r = params.inspiratory_expiratory_ratio; // I:E
-    return seconds(t / (1 + r));
-  }
-
-  const Pressure inspire_pressure_;
-  const Pressure expire_pressure_;
-  Time start_time_;
-  Time inspire_end_;
-  Time expire_end_;
-};
-
-// A discriminated union of "breath finite state machines".
-//
-// TODO(jlebar): Replace with std::variant once we have an STL.  I tried to
-// hack together my own variant class, but it was ultimately disappointing.
-class FsmUnion {
-public:
-  FsmUnion() : mode_(), u_{OffFsm()} {}
-
-  // Replaces the existing breath FSM with a new one, of the given mode and
-  // params.
-  void new_breath(Time now, VentMode m, const VentParams &params) {
-    switch (mode_) {
-    case VentMode_OFF:
-      u_.off.~OffFsm();
-      break;
-    case VentMode_PRESSURE_CONTROL:
-      u_.pressure_control.~PressureControlFsm();
-      break;
-    }
-
-    mode_ = m;
-    switch (mode_) {
-    case VentMode_OFF:
-      new (&u_.off) OffFsm(now, params);
-      break;
-    case VentMode_PRESSURE_CONTROL:
-      new (&u_.pressure_control) PressureControlFsm(now, params);
-      break;
-    }
-  }
-
-  // Gets the state the FSM would like us to (try to) achieve.
-  BlowerSystemState desired_state(Time now) {
-    switch (mode_) {
-    case VentMode_OFF:
-      return u_.off.desired_state(now);
-    case VentMode_PRESSURE_CONTROL:
-      return u_.pressure_control.desired_state(now);
-    }
-    // All cases covered above (and gcc checks this).
-    __builtin_unreachable();
-  }
-
-  // Returns whether or not the current FSM is done with its one breath.
-  bool finished(Time now) {
-    switch (mode_) {
-    case VentMode_OFF:
-      return u_.off.finished(now);
-    case VentMode_PRESSURE_CONTROL:
-      return u_.pressure_control.finished(now);
-    }
-    // All cases covered above (and gcc checks this).
-    __builtin_unreachable();
-  }
-
-private:
-  VentMode mode_;
-  union U {
-    OffFsm off;
-    PressureControlFsm pressure_control;
-  };
-  U u_;
-};
-
-FsmUnion fsm;
-
-} // anonymous namespace
-
-void blower_fsm_init() {}
-
-BlowerSystemState blower_fsm_desired_state(Time now, const VentParams &params) {
+BlowerSystemState BlowerFsm::DesiredState(Time now, const VentParams &params) {
   // Immediately turn off the ventilator if params.mode == OFF; otherwise, wait
   // until the end of a cycle before implementing the mode change.
-  if (params.mode == VentMode_OFF || fsm.finished(now)) {
-    fsm.new_breath(now, params.mode, params);
+  if (params.mode == VentMode_OFF || fsm_.finished(now)) {
```

Comment:
It feels kind of arbitrary that this one member function would be defined in the cc file.  What is the "rule" we should be applying, especially for new people's code?

(Usually putting code into cc files means the build is faster, but like, is our build even running in parallel right now?)

Reply:
The rule I like to follow is just to put nontrivial-sized code in .cpp - however this code is trivial, so I moved it to .h. I wanted to put FsmUnion into .cpp but can't do it cause we need FsmUnion at declaration. Putting it into a separate header feels like overkill.